api/utils/keys/yubikey.go

```diff
@@ -442,59 +473,78 @@ func parsePIVSlot(slotKey uint32) (piv.Slot, error) {
 	}
 }
 
+// selfSignedMetadataCertificate creates a self signed certificate to be stored in the
+// YubiKey's PIV slot. This certificate is purely used as metadata to determine when a
+// slot is in used by a Teleport Client and is not fit to be used in cryptographic operations.
+func selfSignedMetadataCertificate() (*x509.Certificate, error) {
+	// generate a small rsa key to quickly generate a metadata cert.
+	priv, err := rsa.GenerateKey(rand.Reader, 512)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	cert, err := metadataCertificateTemplate()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	cert.PublicKey = priv.Public()
+	if cert.Raw, err = x509.CreateCertificate(rand.Reader, cert, cert, priv.Public(), priv); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return cert, nil
+}
+
 // certOrgName is used to identify Teleport Client self-signed certificates stored in yubiKey PIV slots.
 const certOrgName = "teleport"
 
-func selfSignedTeleportClientCertificate(priv crypto.PrivateKey, pub crypto.PublicKey) (*x509.Certificate, error) {
+func metadataCertificateTemplate() (*x509.Certificate, error) {
 	serialNumberLimit := new(big.Int).Lsh(big.NewInt(1), 128)
 	serialNumber, err := rand.Int(rand.Reader, serialNumberLimit) // see crypto/tls/generate_cert.go
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
 	cert := &x509.Certificate{
 		SerialNumber: serialNumber,
-		PublicKey:    pub,
 		Subject: pkix.Name{
 			Organization:       []string{certOrgName},
 			OrganizationalUnit: []string{api.Version},
 		},
 	}
-	if cert.Raw, err = x509.CreateCertificate(rand.Reader, cert, cert, pub, priv); err != nil {
-		return nil, trace.Wrap(err)
-	}
+
 	return cert, nil
 }
 
-// YubiKeys require touch when generating a private key that requires touch, or using
-// a private key (Sign) with touch required. Unfortunately, there is no good way to
-// check whether touch is cached by the PIV module at a given time. In order to require
-// touch only when needed, we prompt for touch after a short delay when we expect the
-// request would succeed if touch were not required.
+// nonTeleportCertificateError returns a user-readable CompareFailed error.
+// when we run into a non-Teleport certificate, we need to get user input
+// before we overwrite it to avoid breaking their non-Teleport PIV programs.
+func nonTeleportCertificateError(slot piv.Slot, cert *x509.Certificate) error {
+	msg := fmt.Sprintf("Certificate in YubiKey PIV slot %q is not a Teleport client cert:\nSlot %s:\n", slot.String(), slot.String())
+	// Gather a small list of user-readable x509 certificate fields to append to the error message.
+	msg += fmt.Sprintf("\tAlgorithm:\t%v\n", cert.SignatureAlgorithm)
```

Comment:
Please use a strings.Builder or bytes.Buffer. This is too many concats/formatting in a row.